node/pkg/driver/device_connectivity/device_connectivity_iscsi.go

```diff
@@ -32,6 +34,44 @@ func NewOsDeviceConnectivityIscsi(executer executer.ExecuterInterface) OsDeviceC
 	}
 }
 
+func (r OsDeviceConnectivityIscsi) iscsiCmd(args ...string) (string, error) {
+	out, err := exec.Command("iscsiadm", args...).CombinedOutput()
+	return string(out), err
+}
+
+func (r OsDeviceConnectivityIscsi) iscsiDiscoverAndLogin(iqn, ip string) error {
+	logger.Debugf("iscsiDiscoverAndLogin: iqn {%s}, ip {%s}", iqn, ip)
+	output, err := r.iscsiCmd("-m", "discovery", "-t", "sendtargets", "-p", ip)
+	if err != nil {
+		logger.Errorf("Failed to discover iSCSI: {%s}, error: {%s}", output, err)
+		return err
+	}
+
+	output, err = r.iscsiCmd("-m", "node", "-p", ip+":3260", "-T", iqn, "--login")
+	if err != nil {
+		logger.Errorf("Failed to login iSCSI: {%s}, error: {%s}", output, err)
+		return err
+	}
+	return nil
+}
+
+func (r OsDeviceConnectivityIscsi) EnsureLogin(ipsByArrayIdentifier map[string][]string) error {
+	isAnyLoginSucceeded := false
+	var err error
+	for arrayIdentifier, ips := range ipsByArrayIdentifier {
+		for _, ip := range ips {
+			err = r.iscsiDiscoverAndLogin(arrayIdentifier, ip)
+			if err == nil {
+				isAnyLoginSucceeded = true
+			}
+		}
+	}
+	if !isAnyLoginSucceeded {
```

Comment:
True, but I was thinking - if a user decides to manually configure their multipath to allow a single session (per iqn), should we really prevent them?
We already have an indicative message in `MultipleDeviceNotFoundError` later in the code

Reply:
MultipleDeviceNotFoundError returns after rescan, but we should prevent the rescan if it will be failed for sure, because you have to clean up the path.
If you want to support single path, set it as an env.

Reply:
1. Do you mean that the login success count should be per iqn?
2. Could you elaborate about which path do I have to clean up?
3. I wish to allow a user to use a single path, but I don't want to over complicate things for them with an env var.

Reply:
1. not per iqn.
2. If there is only one path, the host os will find one device after rescan, but MultipleDeviceNotFoundError will be returned because it is not a multipath device, so before raising this error to caller,  you should firstly delete this device from host os.
3. ok, so do you have a better idea?

Reply:
1. but I thought the default multipath settings ignores single iscsi paths per iqn (not in total), am I wrong?
2. and what happens if I don't? I assume the user will want it logged in anyway.
3. my idea is to allow both multipath configurations with the same code (without requiring additional actions from the user, without checking their configuration at runtime, etc...), by paying the price of an unnecessary rescan in the rare case of a single iscsi path user that has yet to configure their multipath.conf properly (as we will require them in the user guide).

Reply:
1. do you mean that by default, both iqns of the same svc is being represented by the same multipath device mapper (i.e. if one fails, it will continue to work)?
2. how is it different than never logging out? or never deleting the volume that cannot be acceded (due to login failure)?
3. why? the retry will just re-login it right after. and what if it was already logged it before the driver login?

Reply:
@27149chen can we resolve this? we need this code ASAP

Reply:
Thanks for the approval.